16.0 mig payment slimpay dump restore utils

Honeyxilia opened the pull request.

fcayre committed:
Import the payment_slimpay_dump_restore_utils module into the 12.0 branch

fcayre committed:
Remove spurious recurring_invoices usage from payment_slimpay_dump_restore_utils

legalsylvain committed:
[REF] apply pre-commit, fixing website values

fcayre committed:
Do not activate dump mandates cron by default

fcayre committed:
By default do not restore mandate of partners w/o a contract

fcayre committed:
Strip bic from mandate repr to create

This is an optionnal field (deduced from IBAN by Slimpay), and some value
may be refused (example at time of this writting: CMCIFR2AXXX crashes).

fcayre committed:
Fix exception handling in case of mandate creation crash

fcayre committed:
Skip mandates which partner has been removed from odoo

... instead of crashing.

fcayre committed:
Fix mandates dump util

This is not tested (it should, but not critical) but writing the
resulting json file just crashed in py3k.

legalsylvain committed:
[META] enable flake8

legalsylvain committed:
[META] enable black

legalsylvain committed:
[META] enable isort

legalsylvain committed:
[META] enable black (new version)

Honeyxilia committed:
[IMP] payment_slimpay_dump_restore_utils: black, isort, prettier

Honeyxilia committed:
[MIG] payment_slimpay_dump_restore_utils: Migration to 16.0

Honeyxilia committed:
Adapted to 'account.acquirer' -> 'account.provider' migration

We also renamed variables for consistency.

Honeyxilia committed:
Changed slimpay_client value attributions to function calls

Honeyxilia committed:
Renamed payment.provider file for better indication of inheritance

Honeyxilia committed:
Added explicit contract_payment_auto dependency

The utils method filter_has_contract depends on :
- the contract.contract model, from the OCA module contract.
- the res.partner field payment_token_id, a field declared in contract_payment_auto (dependant on contract)
As such, we add this dependency in order to explicit the requirement.

Honeyxilia committed:
[test] Fixed SlimpayClient import

Honeyxilia committed:
Added an empty dict. return for mandate_doc_to_repr

If the method can't find a partner to link the mandate to, it doesn't return any data, which is then considered a Falsy (None) in get_all_mandates_repr. In order to clarify that possible end point for the method, we add a return with an empty dict, which is also falsy.

Honeyxilia committed:
[test] Added tests for methods used in payment.provider model

These tests methods test whether the high-level helper methods function correctly.
We unfortunally can't test in a test environnement whether the server response is correct, since we have to mock it. However, with other modifications (the files, the records), we can check if the methods performed their operations correctly.

Honeyxilia committed:
[test] Added tests for low-level helpers

For various edge cases not easily covered by the high-level helpers, we add these low-level tests to more granularly tests the helpers and increase coverage.

fcayre commented on payment_slimpay_dump_restore_utils/models/utils.py:
```diff
@@ -57,6 +57,7 @@ def mandate_doc_to_repr(provider, mandate_doc):
             "subscriber": {"reference": partner.id},
             "signatory": signatory,
         }
+    return {}
```
Petite typo dans le message de commit : "link the mandate too" => "link the mandate to".

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,105 @@
+class TestSlimpayDumpProviderUtils(PaymentSlimpayDumpUtilsCommonTC):
+    "This class serves to test the model methods related to this model"
+
+    dump_fp = "/tmp/test_mandates.json"
```
Use the tmpfile python module instead (cleaner than a hard-coded file path and more portable).

Honeyxilia replied:
Fait c:

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,105 @@
+            mandates = json.load(f)
+            self.assertEqual(len(mandates), 2)
+
+        with self.registry.cursor() as test_cursor:
```
Pourquoi utilises-tu un autre curseur ? Ça mérite a minima un commentaire dans le code.
Tu ne vérifies pas le résultat de la restauration des mandats ?

Honeyxilia replied:
A la fin de la méthode `restore_all_missing_mandates`, il y a un commit de l'environnement utilisé, ce qui perturbe la bonne complétion du test. Ainsi, on utilise un environnement et un curseur tier afin de pouvoir finaliser le test correctement

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,105 @@
+        with open(self.dump_fp) as f:
+            mandates = json.load(f)
+
+        # Since the returned mandate is based on the admin partner, we reconstitue the mandate data.
```
Typo : reconstitue => reconstitute

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,105 @@
+            mandates = json.load(f)
+
+        # Since the returned mandate is based on the admin partner, we reconstitue the mandate data.
+        admin_partner_data = slimpay_utils.subscriber_from_partner(self.admin_partner)
```
Pourquoi réutiliser une méthode que tu testes ici ? J'ai l'impression qu'il vaudrait mieux vérifier des valeurs en dur ici, non ?

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,105 @@
+    def test_restore_various_cases(self):
+        """
+        We check various invalid mandates cases
+        """
```
Mettre la docstring sous forme compact si black est d'accord, ça rend plus compact mais aussi lisible (voire plus).

Honeyxilia replied:
Je dois encore compléter le commentaire, donc il peut être plus long - si le commentaire est assez court, je le metterai sur une seule ligne

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,105 @@
+            open(self.dump_fp, "w"),
+        )
+
+        with self.registry.cursor() as test_cursor:
```
Même question concernant le curseur.

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,105 @@
+
+            self.env["payment.provider"].with_env(env)._slimpay_restore_mandates(
+                mandates_fpath=self.dump_fp,
+            )
```
Je n'ai pas compris ce que tu testais ici : seulement que ça ne plante pas ? Que se passe-t-il alors concernant les mandats insérés ?

Honeyxilia committed:
Removed seemingly useless contract autopay util function

Honeyxilia committed:
[test] Improve cov.: Added no cover flag on MissingError catch

The try/except block tries to catch a MissingError exception, which is thrown in the Odoo code when trying to access a record that's been deleted.

The only accessed records in the method are a partner record and its related payment token record (which can be non-existant). Since the start of the method already verifies whether the partner exists, and writing on a non-existant token is not forbidden, the only way for this exception to be thrown would be if a concurrent writing process were to delete that record after the start of the method, which is tricky to test reliably.
So, we add this flag on this exception.

For the Exception branch, while it's more common to encounter issues (for instance, a ValueError trying to access a undefined dict value), we do not consider it interesting.

Honeyxilia committed:
Removed commit statement at the end of util method

Since this method was usually called through command line, having a commit at the end helped to save directly the changes made. However, this is a bad practice, and through the command line, it's better to commit manually.
So, we remove this statement.

Honeyxilia committed:
[test] Changed string argument style

Honeyxilia committed:
[test] Improve cov.: Added test for mandates dump with refresh but without mandates in it

Honeyxilia commented on payment_slimpay_dump_restore_utils/tests/common.py:
```diff
@@ -0,0 +1,129 @@
+    def _mock_search_mandates(self, mocker, pagination=True):
+        """
+        Mock the GET search-mandates endpoint.
+        The pagination argument allows to create another page of not if needed.
```
typo ("another page of not" -> "another page *or* not")

Honeyxilia replied:
in the end, i reworded the comment

fcayre commented on payment_slimpay_dump_restore_utils/tests/test_provider.py:
```diff
@@ -0,0 +1,180 @@
+            self.assertEqual(len(mandates), 2)
+
+        # Phase 2: we restore the dumped mandates.
+        # For the sake of testing, we declare another search-mandate mock.
```
Je n'ai pas compris l'intérêt ici, peux-tu m'expliquer ?

Honeyxilia replied:
l'intêret de faire une restauration ?

Bien que la restauration des mandats récupérés soit un peu pipot (étant donné qu'il s'agit de notre propre retour), cela permet de lancer la méthode avec le provider établi (augmentation de couverture) c:

fcayre commented:
/ocabot merge patch
